src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -3835,14 +3830,18 @@ class CgroupUtils(object):
             # assigned all of the fully avaiable cores. There still may
             # be cores to assign. When use_hyperthreads is disabled, we
             # assign all the cores here.
-            if needed >= 1:
-                if needed > len(cores):
-                    pbs.logmsg(pbs.EVENT_DEBUG4, '%s: %d ncpus still needed' %
-                               (caller_name(), len(cores) - needed))
-                    return {}
-                else:
-                    corelist = sorted(cores)
-                    assigned['cpuset.cpus'] += corelist[:needed]
+            if needed <= 0:
+                # we're done (will always hold if use_hyperthreads
+                # and ncpus_are_cores are both True)
+                pass
+            elif needed <= len(cores):
+                corelist = sorted(cores)
+                assigned['cpuset.cpus'] += corelist[:needed]
+            else:
+                pbs.logmsg(pbs.EVENT_DEBUG4, '%s: %d ncpus still needed' %
+                           (caller_name(), len(cores) - needed))
```

Comment:
`len(cores) - needed` is negative in this case (needed > len(cores)). Do you mean `needed - len(cores)`?

Reply:
I left the message as-is, but you are right that it is confusing; if you're missing 3 cpus it will print "-3 ncpus still needed".

Original code:
```
            if needed > len(corelist):
                pbs.logmsg(pbs.EVENT_DEBUG4, '%s: %d ncpus still needed' %
                           (caller_name(), len(corelist) - needed))
                return {}
```

So you're right that whoever the original author was probably meant needed - len(cores), or in his case needed - len(corelist).

I'll also change the error messages. "Still needed" doesn't say after what they are 'still' needed.

```
            else:
                pbs.logmsg(pbs.EVENT_DEBUG4, '%s: %d more ncpus needed than available' %
                           (caller_name(), needed - len(cores))
                return {}
```

Reply:
But it's all a bit irrelevant, since at the top of the function we have:
```
            avail = len(cores)
            needed = int(requested['ncpus'])
            if self.cfg['use_hyperthreads'] and self.cfg['ncpus_are_cores']:
                needed *= node.cpuinfo['hyperthreads_per_core']
            if needed > avail:
                pbs.logmsg(pbs.EVENT_DEBUG4, 
                           '%s: Insufficient ncpus: %s needed, %s available'
                           % (caller_name(), needed, avail))
                return {}
```
The error message "should never" be printed, unless there is a logic error in the code in between these two.

Checking the code, "avail" is len(cores). In the block between the first and second check on ncpus available, we decrement needed whenever we remove an element in cores. So if the first check passes, it is _impossible_  for the second check to fail, unless someone modifies the code. [I would leave second the guard there -- we definitely do not want to _silently_ assign too few ncpus if we make an error in modified code between the two checks].

Reply:
Note that the first message _is_ actually printed in valid circumstances: if vnode_per_numa_node is False, then jobs are tentatively placed on each socket first, and only if they cannot be fit on a single socket placed on the entire host.